### PR TITLE
DOCSP-7301: Add semantic parse layer

### PR DESCRIPTION
**Pull request by sophstad**
[[JIRA](https://jira.mongodb.org/browse/DOCSP-7301)] [[Spec](https://docs.google.com/document/d/1VPuwEADu4wJtGo4TfA7HfTxBnyy36b3JXGXL_So86ek/edit#heading=h.k0hwkqiwp683)] Interface for adding semantic parse operations to Snooty.

Would appreciate any thoughts on adding tests/test infrastructure since no semantic parse methods exist yet!

### CODE REVIEW

File: snooty/main.py [hunk 1]
```diff
@@ -126,6 +131,14 @@ def on_update(self, prefix: List[str], page_id: FileId, page: Page) -> None:
                 upsert=True,
             )
 
+    def on_update_metadata(
+        self, prefix: List[str], field: Dict[str, SerializableType]
+    ) -> None:
+        page_id = "/".join(prefix)
```
Comment (i80and) [inline snooty/main.py]: I think this variable should be called `property_name` or something

File: snooty/main.py [hunk 2]
```diff
@@ -126,6 +131,14 @@ def on_update(self, prefix: List[str], page_id: FileId, page: Page) -> None:
...
+        self.client["snooty"]["metadata"].update_one(
```
[the remark this answers is not in the document]
Comment (sophstad) [reply snooty/main.py]: Yes, I've implemented this thinking that `field` will be a dictionary that includes our different semantic parse operations as key-value pairs. The dictionaries returned by each independent operation are merged by [this call to `document.update()`](https://github.com/mongodb/snooty-parser/pull/43/files/fe75f5b941ba6e39ac8dab7b9f2a3b8b4be9ec2d#diff-b733f3e75192293503fa7b55bbc62464R19). E.g. `field` may look like this:
```python
{
    slug_title_mapping: {
        server/install: "Install MongoDB",
        server/delete: "Delete Documents from MongoDB"
    },
    toctree: {
        [
            # toctree contents here
        ]
    }
}
```
Therefore the `$set` operation will replace any existing fields and update those that already exist. Perhaps `field` is a bad choice of variable name. Does that make sense?

File: snooty/semanticparser.py [hunk 3]
```diff
@@ -0,0 +1,20 @@
+from typing import Callable, Dict, List
+
+from .types import FileId, Page, SerializableType
+
+
+class SemanticParser:
```
[the remark this answers is not in the document]
Comment (sophstad) [reply snooty/semanticparser.py]: I was thinking class method! Does that make sense to you?

File: snooty/main.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (i80and) [reply snooty/main.py]: :+1:

File: snooty/semanticparser.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (i80and) [reply snooty/semanticparser.py]: Class method is good :+1:

**Pull request merged**